File: frontend/src/types/recommendation.ts
export interface RecommendationRequest {
  average_order_value: number;
  fulfilment_cost_per_purchase: number;
  transaction_cost_per_purchase: number;
  fixed_campaign_operating_cost: number;
  probability_threshold?: number | null;
  filter_campaign?: string | null;
  filter_age?: string | null;
  filter_gender?: string | null;
  filter_interest?: string | null;
}

export interface SegmentRecommendation {
  ad_id: number;
  campaign_id: number;
  age: string;
  gender: string;
  interest: number;
  impressions: number;
  clicks: number;
  spend: number;
  actual_purchases: number;
  conversion_probability: number;
  predicted_class: number;
  ctr: number | null;
  cpc: number | null;
  cac: number | null;
  purchase_conversion_rate: number | null;
  estimated_revenue: number;
  estimated_profit: number;
  recommendation: string;
  recommendation_reason: string;
}

export interface ExecutiveSummary {
  best_campaign_by_profit: string | null;
  lowest_cac_campaign: string | null;
  best_age_group: string | null;
  best_interest_group: string | null;
  largest_inefficient_spend_area: string | null;
  segments_by_recommendation: Record<string, number>;
  estimated_profit_selected: number;
  main_limitation: string;
  recommended_next_action: string;
}

export interface RecommendationRule {
  recommendation: string;
  conditions: string[];
}

export interface RecommendationResponse {
  segments: SegmentRecommendation[];
  executive_summary: ExecutiveSummary;
  rules: RecommendationRule[];
  threshold_used: number;
  total_segments: number;
  limitations: string[];
}

export type RecommendationLabel =
  | "Increase budget carefully"
  | "Continue"
  | "Monitor"
  | "Reduce budget"
  | "Pause";

File: backend/app/schemas/recommendation.py
from typing import List, Optional

from pydantic import BaseModel, Field


class RecommendationRequest(BaseModel):
    average_order_value: float = Field(75, ge=0)
    fulfilment_cost_per_purchase: float = Field(35, ge=0)
    transaction_cost_per_purchase: float = Field(2, ge=0)
    fixed_campaign_operating_cost: float = Field(0, ge=0)
    probability_threshold: Optional[float] = Field(None, ge=0, le=1)
    filter_campaign: Optional[str] = None
    filter_age: Optional[str] = None
    filter_gender: Optional[str] = None
    filter_interest: Optional[str] = None


class SegmentRecommendation(BaseModel):
    ad_id: int
    campaign_id: int
    age: str
    gender: str
    interest: int
    impressions: int
    clicks: int
    spend: float
    actual_purchases: int
    conversion_probability: float
    predicted_class: int
    ctr: Optional[float]
    cpc: Optional[float]
    cac: Optional[float]
    purchase_conversion_rate: Optional[float]
    estimated_revenue: float
    estimated_profit: float
    recommendation: str
    recommendation_reason: str


class ExecutiveSummary(BaseModel):
    best_campaign_by_profit: Optional[str]
    lowest_cac_campaign: Optional[str]
    best_age_group: Optional[str]
    best_interest_group: Optional[str]
    largest_inefficient_spend_area: Optional[str]
    segments_by_recommendation: dict[str, int]
    estimated_profit_selected: float
    main_limitation: str
    recommended_next_action: str


class RecommendationRule(BaseModel):
    recommendation: str
    conditions: List[str]


class RecommendationResponse(BaseModel):
    segments: List[SegmentRecommendation]
    executive_summary: ExecutiveSummary
    rules: List[RecommendationRule]
    threshold_used: float
    total_segments: int
    limitations: List[str]
